src/gss_names.c

```diff
@@ -281,6 +281,80 @@ uint32_t gssntlm_import_name(uint32_t *minor_status,
                                        output_name);
 }
 
+size_t gssntlm_get_attrs_count(const struct gssntlm_name_attribute *attrs)
+{
+    size_t attrs_count = 0;
+    if (!attrs) return 0;
+    while (attrs[attrs_count].attr_name != NULL) attrs_count++;
+
+    return attrs_count;
+}
+
+int gssntlm_copy_attrs(const struct gssntlm_name_attribute *src,
+                       struct gssntlm_name_attribute **dst)
+{
+    struct gssntlm_name_attribute *copied_attrs = NULL;
+
+    size_t attrs_count = gssntlm_get_attrs_count(src);
+    if (attrs_count == 0) {
+        *dst = NULL;
+        return 0;
+    }
+
+    copied_attrs = malloc(sizeof(struct gssntlm_name_attribute) *
+                          (attrs_count + 1)); /* +1 for terminator entry */
+    if (copied_attrs == NULL) {
+        return ENOMEM;
+    }
+
+    for (size_t i = 0; i < attrs_count; i++) {
+        /* read-only persistent data ptr can be just copied */
+        copied_attrs[i].attr_name = src[i].attr_name;
+        copied_attrs[i].attr_value.length = src[i].attr_value.length;
+        copied_attrs[i].attr_value.value = malloc(src[i].attr_value.length);
+        if (copied_attrs[i].attr_value.value == NULL) {
+            /* release all previously allocated memory */
+            while (i > 0) {
+                free(copied_attrs[--i].attr_value.value);
+            }
+            free(copied_attrs);
+            return ENOMEM;
+        }
+        memcpy(copied_attrs[i].attr_value.value, src[i].attr_value.value,
+               src[i].attr_value.length);
+    }
+
+    *dst = copied_attrs;
+    return 0;
+}
+
+/* Returns index of attribute in the array or -1 if not found */
+int gssntlm_find_attr(const struct gssntlm_name_attribute *attrs,
+                      const char *attr_name, size_t attr_name_len)
+{
+    int i = 0;
+    if (!attrs) return -1;
+    while (attrs[i].attr_name != NULL) {
+        if (strncasecmp(attrs[i].attr_name, attr_name, attr_name_len) == 0) {
```

Comment:
Why case insensitve comparisons ?
Also these are all zero terminated strings and you do not need to test a substring so a normal strcmp() is better.
a strNcmp() here will match also attributes that have a common attribute name as a substring like: urn:my:foo and urn:my:foobar so it is not ok

Reply:
Case-insensitive comparison via strncasecmp with STRLENOF(body) is done for attributes in Kerberos MIT library (see src/lib/krb5/krb/pac.c, mspac_attr2type). I haven't read RFC too much, have no idea if attributes are case-sensitive and so I just reused this concept as is. 
As for string zero terminator - gss_buffer_t is defined as BINARY buffer (with 'length' field instead of terminator) and I'm afraid that it is not obligated to be zero-terminated at all (this is actually not a real C string).
As the fact, in my code I do use a call to gss_get_name_attribute with urn of size matching to only a body - without zero terminator - and it works fine with MIT library. 

If you know that this is wrong (do we have this defined by RFC? where), please tell so we can file a bug to MIT.

As for matching prefix with more longer attribute name - yes, my bad, I need to match the length of attribute too. Will fix it...

Reply:
So for case-sensitivity, it depends on what is used, URIs often end up being practically case-insensitive because all of their components are, which is why I guess the example you see with the pac in krb5 ends up using case insensitive comparisons.

We technically though do not need to use a URI actually, as RFC6680 provides another way to name things separated by a space, so we could have a name of "gssntlmssp sids" for example, and that could be definied as case-insensitive as it is not a uri form.

I think I would prefer that at this point given we are not going to register a URI with IANA anyway for this work.

In any case we need to check that the lengths match, as we do not want to match against a substring by mistake. So if you need to use strNcmp you first need to compare the lengths of the two strings and bail out if they differ.

This is the check we need:
`if ((len1 == len2) && (strncmp(...) == 0))`

Reply:
I see you already check the lengths in the updated code, thanks.
The only thing tat remains to be decide is whether we want to use a mocked up but not IANA registered urn: or just use a "local" value of "gssntlmssp sids"

Reply:
Ok, for my side urn name is not so important.
Please decide internally and make a required change if needed (you can modify my PR I think).